astropy/table/table.py

```diff
@@ -3644,12 +3648,33 @@ def _convert_col_for_table(self, col):
             qcol = q_cls(col.data, col.unit, copy=False)
             qcol.info = col.info
             qcol.info.indices = col.info.indices
+            if np.issubdtype(getattr(col, 'dtype', None), np.int):
+                self._dtype_before_conversion = col.dtype
```

Comment:
I think logically the warning should be given here. Is there a particular reason not to do that?

It is risky to use some outside attribute to communicate between methods, so would be good to avoid it!

Reply:
I apologize for the long delay.

Using the attribute is indeed hacky, the part that I am most uneasy about in this PR.

The reason that a warning cannot be issued in here in `_convert_col_for_table()` is that it is used in various other scenarios, in addition to the use case conversion in unit assignment `tab['a'].unit = 'km'` this PR tried to address.

E.g., if a warning is issued here, the following codes will also emit warning, even though from an user's perspective, there is no conversion taken place. `t['b']` is consistently a float-based Quantity. 

```python
t = QTable([[1, 2], [3, 4]], names=['a', 'b'])
t['b'] = Column([4, 5], unit='cm')
```

Any idea on how to address this?